pkg/cmd/workflow/run/run.go

```diff
@@ -0,0 +1,382 @@
+package run
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"reflect"
+	"sort"
+	"strings"
+
+	"github.com/AlecAivazis/survey/v2"
+	"github.com/MakeNowJust/heredoc"
+	"github.com/cli/cli/api"
+	"github.com/cli/cli/internal/ghrepo"
+	"github.com/cli/cli/pkg/cmd/workflow/shared"
+	"github.com/cli/cli/pkg/cmdutil"
+	"github.com/cli/cli/pkg/iostreams"
+	"github.com/cli/cli/pkg/prompt"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"gopkg.in/yaml.v3"
+)
+
+type RunOptions struct {
+	HttpClient func() (*http.Client, error)
+	IO         *iostreams.IOStreams
+	BaseRepo   func() (ghrepo.Interface, error)
+
+	Selector string
+	Ref      string
+	JSON     string
+
+	InputArgs []string
+
+	Prompt bool
+}
+
+func NewCmdRun(f *cmdutil.Factory, runF func(*RunOptions) error) *cobra.Command {
+	opts := &RunOptions{
+		IO:         f.IOStreams,
+		HttpClient: f.HttpClient,
+	}
+
+	cmd := &cobra.Command{
+		Use:   "run [<workflow ID> | <workflow name>]",
+		Short: "Create a dispatch event for a workflow, starting a run",
+		Long: heredoc.Doc(`
+			Create a workflow_dispatch event for a given workflow.
+
+			This command will trigger GitHub Actions to run a given workflow file.  The given workflow file must
+			support a workflow_dispatch 'on' trigger in order to be run in this way.
+
+			If the workflow file supports inputs, they can be specified in a few ways:
+
+			- Interactively
+			- As command line arguments specified after --
+			- As JSON, via STDIN
+			- As JSON, via --json
+    `),
+		Example: heredoc.Doc(`
+			# Have gh prompt you for what workflow you'd like to run
+			$ gh workflow run
+
+			# Run the workflow file 'triage.yml' at the remote's default branch, interactively providing inputs
+			$ gh workflow run triage.yml
+
+			# Run the workflow file 'triage.yml' at a specified ref, interactively providing inputs
+			$ gh workflow run triage.yml --ref=myBranch
+
+			# Run the workflow file 'triage.yml' with command line inputs
+			$ gh workflow run triage.yml -- --name=scully --greeting=hello
+
+			# Run the workflow file 'triage.yml' with JSON via STDIN
+			$ echo '{"name":"scully", "greeting":"hello"}' | gh workflow run triage.yml
+
+			# Run the workflow file 'triage.yml' with JSON via --json
+			$ gh workflow run triage.yml --json='{"name":"scully", "greeting":"hello"}'
```

Comment:
I like the ability to pass JSON via stdin! But then I don't think that we need an explicit JSON flag. Since JSON is machine data and likely will be constructed by tools rather than by hand, we should assume that it can be large. Passing potentially large datasets via argument should generally be discouraged since it can trigger the OS maximum limit for process arguments.